cmake/Protobuf.cmake

```diff
@@ -13,30 +13,75 @@ macro(p4c_obtain_protobuf)
     endif()
     find_package(Protobuf CONFIG)
     if(NOT Protobuf_FOUND)
-        find_package(Protobuf REQUIRED)
+      find_package(Protobuf REQUIRED)
     endif()
+    # Backwards compatibility
+    # Define camel case versions of input variables
+    foreach(UPPER
+        PROTOBUF_SRC_ROOT_FOLDER
+        PROTOBUF_IMPORT_DIRS
+        PROTOBUF_DEBUG
+        PROTOBUF_LIBRARY
+        PROTOBUF_PROTOC_LIBRARY
+        PROTOBUF_INCLUDE_DIR
+        PROTOBUF_PROTOC_EXECUTABLE
+        PROTOBUF_LIBRARY_DEBUG
+        PROTOBUF_PROTOC_LIBRARY_DEBUG
+        PROTOBUF_LITE_LIBRARY
+        PROTOBUF_LITE_LIBRARY_DEBUG
+        )
+        if (DEFINED ${UPPER})
+            string(REPLACE "PROTOBUF_" "Protobuf_" Camel ${UPPER})
+            if (NOT DEFINED ${Camel})
+                set(${Camel} ${${UPPER}})
+            endif()
+        endif()
+    endforeach()
+    # Protobuf sets the protoc binary to a generator expression, which are highly problematic.
+    # They are problematic because they are basically evaluated at build time.
+    # However, we may have scripts that depend on the actual build time during the configure phase.
+    # Hard code a custom binary instead.
+    find_program(PROTOC_BINARY protoc)
+
     if(ENABLE_PROTOBUF_STATIC)
       set(CMAKE_FIND_LIBRARY_SUFFIXES ${SAVED_CMAKE_FIND_LIBRARY_SUFFIXES})
     endif()
-    set(PROTOBUF_LIBRARY "protobuf::libprotobuf")
-    set(PROTOBUF_PROTOC_EXECUTABLE "protobuf::protoc")
   else()
     # Google introduced another breaking change with protobuf 22.x by adding abseil as a new
     # dependency. https://protobuf.dev/news/2022-08-03/#abseil-dep We do not want abseil, so we stay
     # with 21.x for now.
     set(P4C_PROTOBUF_VERSION "21.12")
     message("Fetching Protobuf version ${P4C_PROTOBUF_VERSION} for P4C...")
 
+    # Unity builds do not work for Protobuf...
+    set(CMAKE_UNITY_BUILD_PREV ${CMAKE_UNITY_BUILD})
+    set(CMAKE_UNITY_BUILD OFF)
     # Print out download state while setting up Protobuf.
     set(FETCHCONTENT_QUIET_PREV ${FETCHCONTENT_QUIET})
     set(FETCHCONTENT_QUIET OFF)
+    # Build Protobuf with position-independent code.
+    set(CMAKE_POSITION_INDEPENDENT_CODE_PREV ${CMAKE_POSITION_INDEPENDENT_CODE})
+    set(CMAKE_POSITION_INDEPENDENT_CODE ON)
+
     set(protobuf_BUILD_TESTS OFF CACHE BOOL "Build tests.")
-    set(protobuf_BUILD_PROTOC_BINARIES OFF CACHE BOOL "Build libprotoc and protoc compiler.")
+    set(protobuf_BUILD_PROTOC_BINARIES ON CACHE BOOL "Build libprotoc and protoc compiler.")
```

Comment:
We build protoc from scratch, because can interact badly with protoc plugins such as `grpc_cpp_plugin`.